### PR TITLE
Rework object locking

### PR DESCRIPTION
**Pull request by matthiasbeyer**
This PR reworks the object locking.

It
- Adds note why locking is not required in some cases
- Adds locking where it is missing
- Adds ref counting where it is missing

This is **WIP**, the following modules are on my todo-list:
- [X] `src/objects/`
- [ ] `src/compositor/`
- [x] `src/action`
- [x] `src/serialize`
- [x] `src/values`

Feel free to propose more, participate in this PR (just notify me which stuff
you want to do) or the like!

### CODE REVIEW

**Comment by matthiasbeyer** [comment]
Targets #451

**Comment by neithernut** [comment]
You forgot to rebase this one...

**Comment by matthiasbeyer** [comment]
I'm not sure whether I am able to review the compositor code that easily. I would recommend we merge this now and review the compositor code in another step.

File: src/objects/named.c [hunk 1]
```diff
@@ -88,6 +88,8 @@ ws_named_init(
         self->val = getref(val);
     }
 
+    getref(self);
```
Comment (neithernut) [inline src/objects/named.c]: Doesn't make sense.

File: src/objects/object.c [hunk 2]
```diff
@@ -506,21 +506,38 @@ ws_object_cmp(
...
     }
 
-    return type->cmp_callback(o1, o2);
+    res = type->cmp_callback(o1, o2);
+
+out:
+    ws_object_unlock((struct ws_object*) o2); // unlock, remove `const`
+    ws_object_unlock((struct ws_object*) o1); // unlock, remove `const`
+    return res;
 }
```
Comment (neithernut) [inline src/objects/object.c:553]: This will cause deadlocks.

Comment (matthiasbeyer) [inline src/objects/object.c:553]: How to resolve it?

Comment (neithernut) [inline src/objects/object.c:553]: Loop with try-locks.

Comment (matthiasbeyer) [inline src/objects/object.c:553]: This could be done on the inner (the second) lock only, right?

Comment (neithernut) [inline src/objects/object.c:553]: Not quite, you also have to release the outer lock, if the inner one fails.

Comment (neithernut) [inline src/objects/object.c:553]: You should still loop instead of returning `-EAGAIN`.

Comment (matthiasbeyer) [inline src/objects/object.c:553]: This was done in the last push, the diff is outdated.

**Comment by neithernut** [comment]
ACK for squashing.

**Comment by matthiasbeyer** [comment]
Done with squashing.

**Comment by neithernut** [comment]
ACK.

**Pull request merged**